var/spack/repos/builtin/packages/py-fparser/package.py

```diff
@@ -0,0 +1,58 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#
+from spack import *
+
+
+class PyFparser(PythonPackage):
+    """Parser for Fortran 77..2003 code."""
+
+    homepage = "https://github.com/stfc/fparser"
+    url      = "https://github.com/stfc/fparser/archive/0.0.5.tar.gz"
+    giturl   = "https://github.com/stfc/fparser.git"
+
+    version('0.0.6', git=giturl,
+            commit='638c51ec57cf17624505b70321c3784e356b8910')
+    version('0.0.5', git=giturl,
+            commit='a3ff86b635f7bd7bd281ee94cbd3d9455b288fd9')
+    version('develop', git=giturl, branch='master')
+
+    depends_on('py-setuptools', type='build')
+
+    depends_on('py-numpy', type=('build', 'run'), when='@0:0.0.5')
+    depends_on('py-nose', type='build')
```

Comment:
Are you sure `py-nose` isn't a test dependency as well?

Reply:
`py-nose` is definitely needed during the build, because it is mentioned as a requirement in setup.py. The tests still fail with the same error if I use:

```
    depends_on('py-setuptools', type=('build', 'run', 'test'))
    depends_on('py-nose', type=('build', 'run', 'test'))
    depends_on('py-six', type=('build', 'run', 'test'))
    depends_on('py-pytest', type='test')
```

The only way to make the tests work is to use `type='build'` or `type='run'` in `depends_on('py-pytest', ...)`.

Reply:
Sorry, `type='run'` doesn't work on its own for py-pytest. But it does work with `type=('test', 'run')` or `type='build'`. Does that suggest anything to you?

Reply:
@scheibelp knows a lot more than I do about how `type='test'` is implemented, so I'll leave that question for him.

Reply:
FYI: I have found a bug where some setup is accidentally omitted for test dependencies. I'll be submitting a PR shortly.

Regarding intended usage of the test deptype:

In short, setting `depends_on('py-pytest', deptype='test')` should be sufficient and you should not have to add `test` to the deptype of other dependencies.

If a package has the `build` deptype it does not need the `test` deptype: the `test` deptype indicates a dependency that is only needed for testing (so it is not built at all unless you do `spack install --test=...`). All environment modifications that occur for the `test` deptype also occur for the `build` deptype, so if you have a package that is needed for building and also supplies testing infrastructure, you do not have to set `deptype=('build', 'test')`.

Reply:
> `py-nose` is definitely needed during the build, because it is mentioned as a requirement in setup.py.

I think this is actually a bug on their part. They might not know how to add `test` dependencies to setuptools. The only places it is used is in their unit tests. But let's keep it there for now, otherwise setuptools will try to install it for us.